docs/proposals/component-registration.md

```diff
@@ -0,0 +1,320 @@
+
+
+## Use Cases
+
+This proposal primarily covers features that can be achieved internally by Kubernetes itself, but many of these features and changes are motivated by external use cases, which need enablement. Not all of these use cases will be completely enabled by this proposal.
+
+
+### Deployment
+
+As a k8s operator, I want a to be able to add/remove new components (instances or types) to a new or running cluster (ideally without manually registering it with the apiserver).
+
+As a k8s operator, I want a kubectl command that allows me to validate that the cluster has been deployed and is fully functional.
+
+As a k8s operator, I want a kubectl command that allows me to check which components are deployed/installed on a specific cluster in order to infer which features it supports.
+
+Note: Inferring features from components isn't very useful with vanilla k8s, but becomes more useful as 3rd party or optional components are added (e.g. OpenShift) and/or if components-controller is split into multiple components.
+
+### Debugging
+
+As a k8s operator, I want a single endpoint (and kubectl command) where I can see the current condition of the cluster so that I can debug observed failures (dropped connections, unresponsiveness, etc.).
+
+Note: The endpoint should contain cause where possible, but at least the current condition of the components.
+
+### Notification
+
+As a k8s operator, I want to be notified when the cluster requires human intervention so that I can intervene.
+
+As an external notification system (e.g. PagerDuty), I want an endpoint that responds to regular requests with the current cluster condition so that I can determine (with my own rule set) when to notify the k8s operator.
+
+Note: What condition requires human intervention is not a rule set that can be hard-coded, is likely to be different for each cloud platform, changes based on component configuration, and is likely to require a time-series database.
+
+### Reporting
+
+As a k8s operator, I want to see a report/graph of the readiness/uptime/availability of the cluster over time so that I can make/satisfy SLAs.
+
+Note: At least requires a time-series database and probably a reporting engine.
+
+### Self-Healing
+
+As an k8s operator, I want the cluster to self-heal when individual components crash so that I don't get notified or need to intervene.
+
+As a local process monitoring tool (e.g. monit), I want to be able to determine the condition of the component process I am monitoring to determine if it needs to be restarted.
+
+As a remote container/vm monitoring tool (e.g. bosh monitor), I want to be able to determine the health of the component container/vm I am monitoring.
+
+Note: Self-healing requires knowing how to deploy the cluster components, which is different for each cloud platform and changes based on component configuration.
+
```

Comment:
Missing some use cases here - would recommend adding the section below for Discovery.  They are not quite orthogonal to this solution because any design that identifies services for admins should be able to identify services for clients.

```
## Discovery

As a cluster administrator, I want to make certain ingress points (services with nodeports or loadbalancers, new ingress object) visible as API endpoints offered to the clients of the cluster.  

As a client, I want to get a list of the API groups (Kube, custom, and third party extended) that I should use to perform actions against the cluster - examples are DNS, logging, metrics, or cluster backup.